content/posts/2018-03-18-dependency-upgrade.md

```diff
@@ -0,0 +1,145 @@
+This amount of changes will affect everyone on your team, especially if you use RxJava heavily.
+
+Retrofit might be simpler in that regard even if [changes are not so small](https://github.com/square/retrofit/blob/master/CHANGELOG.md#version-200-2016-03-11).
+The reason is simple — scoping. Retrofit affects your network layer, but not every developer
+actually needs to know how network calls work if you have proper abstractions in place.
+
+# Brace for Impact
+
+Take a step back. Look at the bigger picture. Do you see some patterns here and there? Good.
+
+The thing is, some preparation actions can be done beforehand. For example, RxJava 2.+
+does not allow `null` values in streams. You already can refactor the code to prepare for that,
+most likely using [some kind of `Optional` values](https://github.com/gojuno/koptional).
+
+Using the Retrofit example `RequestInterceptor` was replaced with OkHttp `Interceptor`.
+It is possible to do the refactoring using the Retrofit 1.+ doing no harm at all.
+
+Be pragmatic, it helps! Like, in life!
+
+# Don’t be a Hero
+
+The obvious approach is to make a huge refactoring, but if you have such thoughts
+better take a deep breath and save your soul before it’s too late.
+
+* Code review of a huge refactoring is essentially a huge Skip button.
+  No one on the planet is capable of thinking clearly about something affecting basically everything.
+* Since some non-deterministic conditions can lead to runtime errors
+  a global refactoring is far too dangerous to do in one sitting.
+  Plus your QA department (if you have one) will not be happy to do a full regression.
+  Since it is not focused some edge cases might become broken.
+
+Do the migration gradually. Even better — pick the area of the project with the least impact
+and do experiments there. Is is extremely trivial to do with RxJava. Pretty much
+the same can be done with Retrofit — just move a subset of API declaration
+to another `interface` and you are ready to go. Live with the migrated subset
+for a couple of releases, take a look at metrics (the most trivial one is crash rate)
+and refine your approach. This is not a sprint but a marathon. You should have
+a single goal in mind — maintain the product quality at all costs.
+
+The gradual migration requires different artifacts and package names to avoid conflicts.
+[Jake has this topic covered pretty well](http://jakewharton.com/java-interoperability-policy-for-major-version-updates/).
+Both RxJava 2.+ and Retrofit 2.+ actually apply this policy, allowing using two versions of a library
+in parallel. Some libraries do not follow it though. In such cases, I suggest to repackage
+the previous version using a custom package name and publish a local artifact. This way you can just remove
+the old artifact after the migration is done.
+
+Another word of advice related to gradual migration is following the hierarchy from roots to the upper level.
```

Comment:
Add a subtitle for this paragraph? Maybe even split into multiple paragraphs for better readability